### PR TITLE
Add scripts to make version and deploy to nexus easy

### PR DESCRIPTION
**Pull request by guptasu**

### CODE REVIEW

File: utils/after_success.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# This script is meant to be used by Travis-CI to publish artifacts (binary, sorce and javadoc jars) when releasing snapshots.
+# Right now it will be manually triggered to push jars to nexus server.
+
+SITE_VERSION=$(./gradlew properties | grep -o 'version.*' | sed -n -e 's/^.*version: //p')
+if [ "$SITE_VERSION" == "" ]; then
+    echo "Could not determine the version, so we're exiting."
+    exit 1
+fi
+if [ "${SITE_VERSION##*-}" != "SNAPSHOT" ]; then
+    URL=http://104.197.230.53:8081/nexus/content/repositories/snapshots/io/gapi/gapi-tools-framework/$SITE_VERSION/
```
Comment (scottbla) [inline utils/after_success.sh]: Where did that URL come from?

File: utils/after_success.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+# This script is meant to be used by Travis-CI to publish artifacts (binary, sorce and javadoc jars) when releasing snapshots.
```
Comment (scottbla) [inline utils/after_success.sh]: "sorce" -> "source"

File: utils/after_success.sh (hunk 1 above)
Comment (guptasu) [inline utils/after_success.sh]: This is where we deploy the jars to nexus server.

File: utils/after_success.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+# This script is meant to be used by Travis-CI to publish artifacts (binary, source and javadoc jars) when releasing snapshots.
+# Right now it will be manually triggered to push jars to nexus server.
+
+SITE_VERSION=$(./gradlew properties | grep -o 'version.*' | sed -n -e 's/^.*version: //p')
```
Comment (garrettjonesgoogle) [inline utils/after_success.sh]: What occurs before `version` on the output of `./gradlew properties`? If it is just whitespace, then use `\s*` instead of `.*`. The `.*` could be too grabby if there are new lines in the future that contain `version`. Add all these things to the first grep pattern too. (If you match just 'version' in the first one, there are cases where the second replacement won't happen.)

File: utils/after_success.sh (hunk 1 above)
Comment (garrettjonesgoogle) [inline utils/after_success.sh]: I'm uncomfortable with the IP address here repeating the one contained in build.gradle.

File: utils/update_doc_version.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+# This script updates the READMEs with the latest non-SNAPSHOT version number.
+# Example: Suppose that before running this script, the build.gradle reads 7.8.9.  This script will replace 
+# all occurrences of #.#.# with 7.8.9 in the README files.
+
+# Get the current maven project version.
```
Comment (garrettjonesgoogle) [inline utils/update_doc_version.sh]: maven? maybe just change this to "current project version."

File: utils/update_doc_version.sh [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+RELEASED_VERSION=$(./gradlew properties | grep -o 'version.*' | sed -n -e 's/^.*version: //p')
+echo "Released version is $RELEASED_VERSION"
+echo "Changing version to $RELEASED_VERSION in README files"
+sed -ri "s/[0-9]+\.[0-9]+\.[0-9]+(-SNAPSHOT)?/$RELEASED_VERSION/g" README.md
```
Comment (garrettjonesgoogle) [inline utils/update_doc_version.sh]: I'm a little concerned about someone referencing a version of something else in the README, and that getting replaced.

File: utils/after_success.sh (hunk 3 above)
Comment (scottbla) [inline utils/after_success.sh]: Actually...the "-o" option to grep tells it to only output the matching part...so its output should either be empty (no "version") or start with "version".  I don't know what gradlew properties looks like, but I would say that the grep is unneeded.  If it outputs one property per line:

```
  ./gradlew properties | sed -n -e 's/^version: *//p'
```

(does sed support the \s sequence for whitespace?)

File: utils/update_doc_version.sh (hunk 5 above)
Comment (scottbla) [inline utils/update_doc_version.sh]: Or if there's an IP address in there for some reason?

Is there a reasonable way for the README.md file to be constructed from some template file instead of doing in-place edits?

File: utils/after_success.sh (hunk 1 above)
Comment (guptasu) [inline utils/after_success.sh]: This IP will eventually be replaced by the proper maven central url. However, I think its important to  guard ourselves from accidentally pushing t release branch multiple times.

File: utils/after_success.sh (hunk 3 above)
Comment (garrettjonesgoogle) [inline utils/after_success.sh]: If there are multiple output lines, then the -o step is necessary, because we don't want the non-version properties.

File: utils/after_success.sh (hunk 1 above)
Comment (garrettjonesgoogle) [inline utils/after_success.sh]: I agree with the logic, I just have an issue with repeating the ip address - I think it should be defined in one place.

File: utils/after_success.sh (hunk 3 above)
Comment (guptasu) [inline utils/after_success.sh]: thanks for the feedback. Seems to work.

File: utils/update_doc_version.sh (hunk 4 above)
Comment (guptasu) [inline utils/update_doc_version.sh]: Done

File: utils/update_doc_version.sh (hunk 5 above)
Comment (guptasu) [inline utils/update_doc_version.sh]: That can be the case and we can fix this script later. Anyways, these things will be caught when someone reviews your PR for README update. We should refine this script then.

Comment (guptasu) [inline utils/update_doc_version.sh]: Or perhaps stop using this utility script and do a manual update.

File: utils/update_doc_version.sh [hunk 6]
```diff
@@ -4,7 +4,7 @@
 # Example: Suppose that before running this script, the build.gradle reads 7.8.9.  This script will replace 
 # all occurrences of #.#.# with 7.8.9 in the README files.
 
-# Get the current maven project version.
+# Get the current project version.
 RELEASED_VERSION=$(./gradlew properties | grep -o 'version.*' | sed -n -e 's/^.*version: //p')
```
Comment (scottbla) [inline utils/update_doc_version.sh]: Maybe that command line should be split into a separate script to be used in all the places it's needed?  Looks like you changed it in update_gradle_version.sh but not here :)  (is the same thing elsewhere?)

Comment (garrettjonesgoogle) [inline utils/update_doc_version.sh]: it could be made into a gradle task, it doesn't need to be a script - it could be called like this:

`./gradlew version`

File: utils/after_success.sh (hunk 3 above)
Comment (scottbla) [inline utils/after_success.sh]: Garrett - the "-n" option to sed tells it to not print lines by default.  Then the "s/...//p" tells sed to print the result of the substitution if it does something.  The net result is that the only lines that sed outputs should be the ones starting with "version:" with the leading "version:\s*" stripped off.

Comment (garrettjonesgoogle) [inline utils/after_success.sh]: oh neat!

File: utils/after_success.sh (hunk 1 above)
Comment (scottbla) [inline utils/after_success.sh]: Maybe a better approach would be to have a single "common_vars.sh" file (or some such) that defines common things like:

   NEXUS_HOST_PORT=104.197.230.53:8081
   CURRENT_VERSION=$(./gradlew properties | sed -n -e 's/^version: *//p')

then at the top of every other *.sh file, add this:

   source ./common_vars.sh

Comment (garrettjonesgoogle) [inline utils/after_success.sh]: The issue with the ip address is that it appears in a file that is not part of the PR, `build.gradle`, which doesn't interact with bash.

Comment (scottbla) [inline utils/after_success.sh]: I'm not sure what you mean there...do you not have access to that file when running these scripts?

Comment (garrettjonesgoogle) [inline utils/after_success.sh]: The problem isn't these scripts, it is that gradle also needs to know the IP address when it does a publish.

Comment (scottbla) [inline utils/after_success.sh]: OK - I know little about gradle/gradlew, so I'll defer to you on that part :)

**Comment by scottbla** [comment]
OK...I think it looks OK to me, but I haven't been having much luck figuring out this review interface or know if there's any buttons I should press to say "OK".

File: utils/after_success.sh [hunk 7]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+# This script is meant to be used by Travis-CI to publish artifacts (binary, source and javadoc jars) when releasing snapshots.
+# Right now it will be manually triggered to push jars to nexus server.
+
+SITE_VERSION=$(./gradlew properties | sed -n -e 's/^\s*version: //p')
+NEXUS_SNAPSHOTS_DIR=$(cat ./build.gradle | sed -ne 's/def\s+privateSnapshotRepo\s*=\s*"\(.*\)".*/\1/p')
```
Comment (garrettjonesgoogle) [inline utils/after_success.sh]: Dependency on this variable seems problematic - someone might rename it and not know that it affects this script.

Comment (guptasu) [inline utils/after_success.sh]: I will think about making this clean and less error prone. Let me know if you have suggestions.

Comment (garrettjonesgoogle) [inline utils/after_success.sh]: Make a gradle task called "printSnapshotRepo" which prints the value you need. If it is a gradle task, people will know that there will be external references to it.
